File: 0556-Next Greater Element III/0556-Next Greater Element III.cpp
class Solution {
public:
    int nextGreaterElement(int n) {
        string str = to_string(n);
        if (str.size() == 1)
        {
            return -1;
        }
        int i = str.size() - 2;
        while (i >= 0 && str[i] >= str[i + 1])
        {
            --i;
        }
        if (i == -1)
        {
            return -1;
        }
        
        int j = str.size() - 1;
        while (j >= i && str[i] >= str[j])
        {
            --j;
        }
        
        swap(str[i], str[j]);
        reverse(str.begin() + i + 1, str.end());
        long long result = stol(str);
        if (result > INT_MAX)
        {
            return -1;
        }
        
        return result;
    }
};

File: 0501-0600/0556-Next Greater Element III/0556-Next Greater Element III.py
class Solution:
    def nextGreaterElement(self, n: int) -> int:
        num = list(str(n))
        i = len(num) - 2
        while i >= 0:
            if num[i] < num[i + 1]:
                break
            i -= 1
        if i == -1:
            return -1
          
        j = len(num) - 1
        while num[j] <= num[i]:
            j -= 1
        num[i], num[j] = num[j], num[i]
        result = int(''.join(num[:i + 1] + num[i + 1:][::-1]))
        return result if result < (1 << 31) else -1
